arch/risc-v/src/mpfs/mpfs_coremmc.c

```diff
@@ -0,0 +1,2406 @@
+                          uint32_t rlong[4]);
+static int  mpfs_recvshort(struct sdio_dev_s *dev, uint32_t cmd,
+                           uint32_t *rshort);
+
+/* Event handler */
+
+static void mpfs_waitenable(struct sdio_dev_s *dev,
+                            sdio_eventset_t eventset, uint32_t timeout);
+static sdio_eventset_t mpfs_eventwait(struct sdio_dev_s *dev);
+static void mpfs_callbackenable(struct sdio_dev_s *dev,
+                                sdio_eventset_t eventset);
+static int  mpfs_registercallback(struct sdio_dev_s *dev,
+                                  worker_t callback, void *arg);
+static void mpfs_callback(void *arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct mpfs_dev_s g_coremmc_dev =
+{
+  .dev =
+  {
+    .reset            = mpfs_reset,
+    .capabilities     = mpfs_capabilities,
+    .status           = mpfs_status,
+    .widebus          = mpfs_widebus,
+    .clock            = mpfs_clock,
+    .attach           = mpfs_attach,
+    .sendcmd          = mpfs_sendcmd,
+    .blocksetup       = mpfs_blocksetup,
+    .recvsetup        = mpfs_recvsetup,
+    .sendsetup        = mpfs_sendsetup,
+    .cancel           = mpfs_cancel,
+    .waitresponse     = mpfs_waitresponse,
+    .recv_r1          = mpfs_recvshortcrc,
+    .recv_r2          = mpfs_recvlong,
+    .recv_r3          = mpfs_recvshort,
+    .recv_r4          = mpfs_recvshort,
+    .recv_r5          = mpfs_recvshortcrc,
+    .recv_r6          = mpfs_recvshortcrc,
+    .recv_r7          = mpfs_recvshort,
+    .waitenable       = mpfs_waitenable,
+    .eventwait        = mpfs_eventwait,
+    .callbackenable   = mpfs_callbackenable,
+    .registercallback = mpfs_registercallback,
+  },
+  .hw_base           = CONFIG_MPFS_COREMMC_BASE,
+  .plic_irq          = MPFS_IRQ_FABRIC_F2H_0 + CONFIG_MPFS_COREMMC_IRQNUM,
+  .blocksize         = 512,
+  .fifo_depth        = 0,
+  .onebit            = false,
+  .polltransfer      = true,
+  .multiblock        = false,
+  .waitsem           = SEM_INITIALIZER(0),
+};
+
+/* Not all requests are 32-bit aligned, use a spare buffer workaround */
+
+static uint32_t g_aligned_buffer[512 / 4];
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mpfs_modifyreg8
+ *
+ * Description:
+ *   Atomically modify the specified bits in a memory mapped register
+ *
+ * Input Parameters:
+ *   addr       - address to use
+ *   clearbits  - bit clear mask
+ *   setbits    - set bit mask
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void mpfs_modifyreg8(uintptr_t addr, uint8_t clearbits,
```

Comment:
Maybe it is time for `void modifyreg8(unsigned int addr, uint8_t clearbits, uint8_t setbits)` interface placed in `arch/risc-v/src/common`? I see `mpfs_modifyreg8` in `arch/risc-v/src/mpfs/mpfs_usb.c` that duplicates 95% of this.

Reply:
Thanks, fixed all the others but will come back to this, as it'd affect multiple drivers / files.

Reply:
Sure. That is more an improvement and should be done in a separate PR.